cpp/src/string/nvcategory_util.cu

```diff
@@ -0,0 +1,170 @@
+			DEVICE_ALLOCATED);
+
+	//This is questionable behavior and should be reviewed by peers
+	//Smart pointers would be lovely here
+	if(column->dtype_info.category != nullptr){
+		NVCategory::destroy(column->dtype_info.category);
+	}
+	column->dtype_info.category = new_category;
+
+	NVStrings::destroy(strings);
+	return GDF_SUCCESS;
+}
+
+gdf_error validate_categories(gdf_column * input_columns[], int num_columns, gdf_size_type & total_count){
+	gdf_dtype column_type = GDF_STRING_CATEGORY;
+	total_count = 0;
+	for (int i = 0; i < num_columns; ++i) {
+		gdf_column* current_column = input_columns[i];
+		if (nullptr == current_column) {
+			return GDF_DATASET_EMPTY;
+		}
+		if ((current_column->size > 0) && (nullptr == current_column->data))
+		{
+			return GDF_DATASET_EMPTY;
+		}
+		if (column_type != current_column->dtype) {
+			return GDF_UNSUPPORTED_DTYPE;
+		}
+		total_count += input_columns[i]->size;
+	}
+	return GDF_SUCCESS;
+}
+
+#include <iostream>
+gdf_error concat_categories(gdf_column * input_columns[],gdf_column * output_column, int num_columns){
+
+	gdf_size_type total_count;
+	gdf_error err = validate_categories(input_columns,num_columns,total_count);
+	if(err != GDF_SUCCESS) return err;
+
+	if(total_count >  output_column->size){
+		return GDF_COLUMN_SIZE_MISMATCH;
+	}
+	if(output_column->dtype != GDF_STRING_CATEGORY){
+		return GDF_UNSUPPORTED_DTYPE;
+	}
+	//TODO: we have no way to jsut copy a category this will fail if someone calls concat
+	//on a single input
+	if(num_columns < 2){
+		return GDF_DATASET_EMPTY;
+	}
+	NVCategory * new_category = input_columns[0]->dtype_info.category;
+	NVCategory * temp_category;
+
+	for (int i = 1; i < num_columns; i++) {
+		NVStrings * temp_strings = input_columns[i]->dtype_info.category->to_strings();
+		temp_category = new_category->add_strings(*temp_strings); //this is the only way to add to a category and keep the dictionary sorted
+		if(i > 1){
+			//only destroy categoryy after first iteration
+			NVCategory::destroy(new_category);
+
+		}
+
+		NVStrings::destroy(temp_strings);
+
+		new_category = temp_category;
+	}
+
+	new_category->get_values(
+			static_cast<nv_category_index_type *>(output_column->data),
+			true);
+	output_column->dtype_info.category = new_category;
+
+
+
+	return GDF_SUCCESS;
+}
+
+gdf_error sync_column_categories(gdf_column * input_columns[],gdf_column * output_columns[], int num_columns){
+
+	if(num_columns == 0){
+		return GDF_DATASET_EMPTY;
+	}
+	gdf_size_type total_count;
+
+	gdf_error err = validate_categories(input_columns,num_columns,total_count);
+	GDF_REQUIRE(GDF_SUCCESS == err, err);
+
+	err = validate_categories(output_columns,num_columns,total_count);
+	GDF_REQUIRE(GDF_SUCCESS == err, err);
+
+	for(int column_index = 0; column_index < num_columns; column_index++){
+		if(input_columns[column_index]->size != output_columns[column_index]->size){
+			return GDF_COLUMN_SIZE_MISMATCH;
+		}
+	}
+
+	NVStrings** temp_strs = new NVStrings*[num_columns];
```

Comment:
Doesn't look like `temp_strs` is ever freed. 

Can you use a `std::vector<NVStrings*>` for this instead?